test/testRules.js

```diff
@@ -1183,6 +1184,74 @@ export const testRules = {
         });
         test.deepEqual(actual, expected);
 
+        test.done();
+    },
+
+    testResourceCompleteness: function(test) {
+        test.expect(5);
+
+        const rule = new ResourceCompleteness();
+        test.ok(rule);
+
+        const subjects = [
+            {},
+            { key: "completeness.test.source-missing", source: undefined },
+            { key: "completeness.test.target-missing", target: undefined },
+            { key: "completeness.test.both-missing", source: undefined, target: undefined },
+        ]
+            .map((overrides) => ({
+                key: "completeness.test",
+                sourceLocale: "en-US",
+                source: "Some source string.",
+                targetLocale: "de-DE",
+                target: "Some target string.",
+                pathName: "completeness-test.xliff",
+                state: "translated",
+                ...overrides, // override some of properties
+            }))
+            .map((props) => new ResourceString(props))
+            .map((resource) => ({ locale: "de-DE", resource, file: "x/y" }));
+
+        const results = subjects.map((subject) => rule.match(subject));
+
+        test.equal(results[0], undefined);
+
+        const commonResultProps = {
+            severity: "error",
+            rule,
+            pathName: "x/y",
+            locale: "de-DE",
+            description: "Resource must have both source and target element defined",
+        };
+        
+        test.deepEqual(
+            results[1],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.source-missing",
+                source: undefined,
+                highlight: "The following elements are missing in the resource: <e0>source</e0>",
+            })
+        );
+        test.deepEqual(
+            results[2],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.target-missing",
+                source: "Some source string.",
+                highlight: "The following elements are missing in the resource: <e0>target</e0>",
+            })
+        );
+        test.deepEqual(
+            results[3],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.both-missing",
+                source: undefined,
+                highlight: "The following elements are missing in the resource: <e0>source, target</e0>",
+            })
+        );
+
```

Comment:
Also, there should be a test with a normal resource that has both source and target where calling match() on it does not produce a Result.

Additionally, we have to think about the case where the linter finds the source en-US.xliff or en-US.properties and there are only source tags in there with no targets. I think that probably should not generate errors if the target tag is missing when the source locale of the resource matches the source locale for the whole project.

Reply:
I'm not sure about that. Given that lack of a target element is technically correct [per XLIFF spec](https://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html#trans-unit:~:text=Zero%20or%20one%20%3Ctarget%3E%20element), when processing an XLIFF document one could reasonably assume that a missing target element means that the translation for a trans-unit has not been provided yet, hence it's not ready for import and an error should be produced. To me, such conclusion makes sense regardless of the languages used - otherwise, why would someone create a trans-unit and not expect a translation for it in the first place? (except when a trans-unit has an attribute `translate=no` but didn't check yet if this is supported in our library).

Reply:
Let's look at the target use cases for this rule:

1. Checking the returned files from our translation vendors. In this case, they are all xliff files and we want to ensure that all of the files they returned to us include a target tag for every trans-unit. If not, they missed one, and we need to alert them to it.
2. Checking files in our git repos. (For later when we do source code.) In this case, we will encounter various resource files (.xliff in iOS/MacOS, .properties in the React apps and Webapp, .pot for python, .xml in Android, .resx in Windoze) that are the source file of our translations and which are either pushed directly into mojito or converted to xliff first by the loctool and then pushed into mojito. In this case, we don't really want to check the source files. It could have a target tag or not. Doesn't matter to us. But, if it is a translated file, we want to make sure that it has a translation for each resource, especially for file formats where the translated file only contains the translations and not the English string. Example: properties files. To build a full Resource instance out of a properties file, you have to read the translated properties file and the English source properties file separately and fuse them together. It would be easy in this case to have a resource file in some language that has missing or extra strings. We want to alert in this case. Missing strings are an error. Extra strings should probably be a warning.

Reply:
BTW - if you are just checking regular xliff files and you don't mind missing target strings for your project, you can turn off this rule.

Reply:
As of now:
@ 1 already supported
@ 2 source there, target missing > error; source missing, target there: warning; both source and target missing > no error
@ original comment: source there, target missing, but source language == target language > no error (e.g. for en-US > en-GB files)